Allow combining `#[cache]` with arguments

csnover opened the pull request:
This is based on the commits in #416, because I did not want to deal with slow tests :-)

I was adapting a parser for a format (wikitext) which has pathological backtracking behaviour which demands caching to avoid quadratic parsing, but also the format is context-sensitive so rules need to receive arguments in order to know which tokens are terminators or not according to the enclosing context.

This patch adds support for combining `#[cache]` with arguments for the specific use case I had, which is passing around a context object with interior mutability ( :scream: ) by reference.

If you have separate suggestions about how to deal with such an unpleasant format, I am all ears :-)

Thanks, and after years, I still really enjoy working with rust-peg every time I have a need for a non-trivial parser!

csnover committed:
Allow combining `#[cache]` with arguments

This is a naïve first attempt at enabling caching with arguments,
supporting both values and ownable references. References are less
efficient because it is extremely difficult to satisfy the
constraint `K: Borrow<Q>` for `HashMap` with a tuple. Most
recommendations seem to be to either do some contortions with trait
objects, or to use the hashbrown crate and implement
`hashbrown::Equivalent`, but I have not been able to successfully
think through how to do this in a way which works with arbitrary
tuples of foreign types. Still, a half-measure which can be
improved later is hopefully better than nothing.

kevinmehall commented on peg-macros/translate.rs:
```diff
@@ -158,11 +158,33 @@ fn make_parse_state(grammar: &Grammar) -> TokenStream {
+                            _ => true,
+                        })
+                        .collect(),
+                    RuleParamTy::Rule(ty) => ty.clone(),
```
I don't think this will work? This is for a parameter declared as `rule<T>`, which at the Rust level is a closure returning `RuleResult<T>`. To work properly, the cache key would need to be the identity of the closure definition site plus any closed-over upvars, which would be quite difficult to do. So this should probably remain forbidden.

csnover replied:
You’re right, I improperly assumed that prohibiting generic types was sufficient for this case because I almost never used `rule` with a concrete type. Now that you point it out, it is embarrassingly obvious that `RuleParamTy` appearing here actually means I needed to think about it more :-)

csnover commented on peg-macros/translate.rs:
```diff
@@ -158,11 +158,33 @@ fn make_parse_state(grammar: &Grammar) -> TokenStream {
+                            // dependency.
+                            TokenTree::Punct(punct) => punct.as_char() != '&',
+                            _ => true,
+                        })
```
What you are saying makes sense.

I approached this largely pragmatically where a move away from “you can’t do this at all” was a step in the right direction, rather than doing the absolute best engineering. (It feels bad to say that when contributing to someone else’s project…) So I considered it OK to just have a compilation error in situations like what you describe with `&str`, though you make a great point that using `<T as ToOwned>::Owned` makes more sense—I always forget about this. (AFAIK it would still require stripping the reference in a hacky way like this, though?)

The issue of value versus reference types in cases similar to this is one which has come up in another crate I maintain and I never really came up with a good solution for what to do in the absence of type metaprogramming, and so I am not sure if there is any general solution for this case either.

I think the main issue with *requiring* cacheable arguments to be passed by value is that cloning is only necessary when inserting to the cache. For rules where there is a cache hit, this would just be wasted work. Additionally, in my case, requiring to pass by value would also require adding `Rc`, whereas right now it is not needed. These are not dealbreakers but it does feel bad to artificially restrict performance in a feature intended to improve performance.

kevinmehall replied:
Oh, I missed the part where you mentioned interior mutability :scream:. In that case, `Rc` doesn't even save you because then you'd be mutating the copy used as the cache HashMap key, which it's not going to like.

I'm kind of inclined to say that rule arguments *should* be simple, immutable, maybe even `Copy` types and anything beyond that is asking for trouble. For example, I think `pub rule`s already require arguments to be `Copy` when they re-parse for error position.

Would it work for your use case to have an escape hatch attribute, maybe `#[cache_ignore]` that excludes an argument from the cache key but passes it into the rule anyway? You would pass your context twice, once immutable by value for the cache key, and once as a mutable reference and `#[cache_ignored]` for the rule to mutate.

csnover replied:
> In that case, Rc doesn't even save you because then you'd be mutating the copy used as the cache HashMap key, which it's not going to like.

Since there are no blanket implementations of `Hash` for interior-mutable types, I have to implement that trait myself, so I think this is no issue, as the `Rc` implementation is just a pass-through, and it really would just be the performance loss?

> Would it work for your use case to have an escape hatch attribute, maybe #[cache_ignore] that excludes an argument from the cache key but passes it into the rule anyway?

I am not sure, I think I would have to run some tests, or wait for a day when I can think better :-). The context would actually be `Copy` except for these damnable interior-mutable properties :-).

To explain more what is going on here in the first place: The original Wikitext parser did not backtrack for mismatched pairs; once it saw `[[ {{` it would only ever look for `}}`. This means the obvious rule for parsing templates…

```peg
rule template()
  = "{{" t:template_content() "}}" { Token::Template(t) }
  / "{{" { Token::Text }
```

…does not work, because for the input "[[ {{ ]]" it would produce `Link { content: Text("{{") }` while the “correct” production is actually `Text("[[ {{ ]]")`.

To address this, whenever a mismatched pair is encountered, the fallback subexpression (`rule broken_<whatever>`) consumes the opening delimiter and then *clears* `ctx.prod_kind`. This causes whichever terminator *had* been set by the parent rule to no longer match in `inline_breaks` (the rule which decides whether an expression is finished), which causes the child to consume it instead of stopping at the terminator, which makes the parent rule fail and match its fallback subexpression, and on and on, until all the previous unclosed delimiters are consumed as text.

So, again using the example "[[ {{ ]]", the order of operation becomes:

|   | rule a | rule b | rule c | result |
|-|-|-|-|-|
| attempt 1 | wikilink | template |             text "]]" | *ERROR* (no "}}") |
| `ctx.prod_kind`: |    `Some(Link)`   |    `Some(Template)` | |
| attempt 2 | wikilink |        broken_template |  text  "{{ ]]" | *ERROR* (no "]]") |
| `ctx.prod_kind`:  |  `Some(Link)` |     `None` | |
| attempt 3 | broken_wikilink | broken_template | text "[[ {{ ]]" | *OK* |
| `ctx.prod_kind`: | None  | None | | |

And the rule looks (approximately) like:

```peg
rule template(ctx: &Context)
  = t:(
      ctx:({ ctx.with_prod_kind(Some(ProdKind::Template)) })
      "{{" t:template_content(&ctx) inline_breaks(&ctx) "}}" { Token::Template(t) }
    / /* broken_template */ "{{" { ctx.prod_kind.set(None); Token::Text }
  )
  { t }
```

It would be great to find another way to handle this but this is what the upstream developers came up with in their own PEG which I have adapted.

Mingun replied:
> The simplest and most consistent way would be to just require `Clone + Hash + 'static`.

It seems to me, that most idiomatic way is to introduce `Cacheable` trait, implement it for types, that you known is safe to cache (maybe via blanket impl), and cache them using methods on that trait.

csnover committed:
Fix new clippy lints

csnover commented:
I’ve pushed an update which:

* Rebases on the current master
* Disallows `rule<_>` caching
* Uses `ToOwned`
* Uses a small `macro_rules` helper to remove references. It is still incompetent, but somewhat less incompetent, and cleaner
* Moves a remark from commit message to comments at the place where it is relevant

I stared into the void for a few hours to see if I could figure out a way to avoid cloning reference arguments for `cache.get` but was not successful. There are some low-complexity solutions when using hashbrown instead of std with fixed length arguments of known types, moderate-complexity solutions for tuples of fixed length with std and trait objects, but nothing I could think through with my current level of brain dysfunction to genericise either approach to N<16 tuples of arbitrary foreign types.

I also notice now that I misread your earlier question about “what about passing it in twice?”—apparently my brain instead substituted “what about always only ever allowing values?”. That would certainly work, though I am struggling to think about where the benefit comes from doing that, are you able to say more about it? My apologies!

csnover committed:
Allow specifying lifetime bounds for the grammar lifetimes

Because the list of bounds is used in multiple places, the grammar
does a little contorting of collecting and then unzipping so that
the code generator can continue to just take cheap slices instead
of needing to map and collect.

In the long term, the grammar should also accept a `where` clause,
but this requires more code for merging clauses together, and it
turns out I do not need this anyway, so any more work is wasted
time for me for now.

csnover committed:
Allow combining `#[cache]` with arguments (attempt 2)

This approach introduces an explicit `Cacheable` trait which must
be implemented by users to allow types to be passed through into
the cache. A hopefully reasonable default implementation is
provided for common primitive and container types.

This approach uses the `hashbrown` crate which exposes a more
generic `Equivalent` trait which avoids having to generate the
cached key just to check if a position is already cached. The
crate is re-exported because users will also need to sometimes
implement `Equivalent`. It is possible that only that trait
should be exported but my general experience as a library
consumer is that I often end up feeling sad when some library
only selectively exposes something from another crate because
then I end up needing to import the dependency myself and keep
the versions in sync.

In order to implement this foreign trait and do the correct
ref-to-owned comparisons, newtypes have to be used instead of
tuples, so the codegen generates those.

Somehow, I am perpetually confused about the correct direction
of the `Equivalent` trait, so it is possible that this
implementation is doing something incredibly stupid or has some
trait bounds which are totally screwed up. I do not know why this
particular thing is so hard for my brain to understand.

Relatedly, the introduction of the `Key` associated type and `key`
function (which are bad names, and which I cannot for the life of
me figure out what word to use) exists to get `&T` (e.g. `&bool`)
to work, because it seems impossible otherwise since auto-deref
will never get from a `&bool` to a `bool`, but I might just be a
complete idiot. I added it in a mad fugue state and that managed
to get things working. More generally, it feels like this
implementation ought to be able to rely both more *and* less on
`hashbrown::Equivalent`. I have low confidence that this is a
totally correct solution to the problem of comparing references.

Because there is currently no metaprogramming for types in Rust,
some extra contortions are done in the code generator to allow
users to pass unlabelled references and then have the right thing
happen, in the form of the hidden `chomp_ref!` macro. (i.e. Even
though the final type must be unambiguous, just emitting `<&Foo
as Cacheable>::Cached` causes a compilation error because
anonymous lifetimes are forbidden here, so there has to be a
valid lifetime.)

The use of an additional hidden lifetime seems to be necessary to
avoid tying the cache to a lifetime that extends beyond the
lifetime of the parse in a way which then makes the lifetimes
of borrows unresolvable. (Using `'input` or `'static` causes
various sadnesses, which are not adequately covered in the test
suite right now, since the sadness happens in the context of
passing references into and out of the grammar, and I have
neglected this in lieu of getting an actual implementation written
and maybe reviewed by someone smarter.)

An older version of hashbrown is used to continue supporting the
existing MSRV of rust-peg.

csnover commented:
I return to this Sisyphean task with what is hopefully a more architecturally sound approach using the suggested explicit `Cacheable` trait and `hashbrown` for the internal hash implementation to eliminate the key cloning by using its `Equivalent` trait.

I also, in what is probably an unnecessary additional change, extended the grammar to allow lifetime bounds to be provided (in a separate commit, which probably should be in its own PR, but I having time management issues).

I tried to offer some rationales for what is going on in the commit, let me know if there are additional questions.

One thing I did not have time to do, which should be done if this patch actually looks sane to anyone else (dubious, my IQ is in free fall), is to make sure that there is a unit test for taking references with lifetimes via `grammar` and returning references to `&'input` in rules, since this is not tested now, and it is very easy to accidentally make changes that appear to not do anything bad but actually extend lifetimes in ways that make it impossible to do either of these things.
